keps/sig-apps/3939-allow-recreation-when-fully-terminated/README.md

```diff
@@ -0,0 +1,699 @@
+// +enum 
+type RecreatePodsWhen string
+const (
+ // This is a field that recreates pods when they are marked as terminating or failed
+ // ie this means that as soon as pods get marked for deletion they will be recreated
+ TerminatingOrFailed RecreatePodsWhen = "TerminatingOrFailed"
+ // Only recreate pods when they are marked as failed.
+ Failed              RecreatePodsWhen = "Failed"
+)
+```
+
+```golang
+type JobSpec struct{
+  ...
+ // RecreatePodsWhen specifies when pods should be recreated.
+ // TerminatingOrFailed means to recreate when a pod is either terminating or failed
+ // TerminatingOrFailed is the default.
+ // Failed means to wait until pods are fully terminated or failed before recreating
+ // +optional
+ RecreatePodsWhen *RecreatePodsWhen
+}
+```
+
+So we can count terminating pods separately from active or failed we need to include a new field in the JobStatus.
+
+```golang
+type JobStatus struct {
+  ...
+  // Number of terminating pods
+  // +optional
+  terminating *int32
+}
+```
+
+We will allow only opt-in behavior for this feature so we will fall back to `TerminatingOrFailed` for `RecreatePodsWhen`.  
+
+### Implementation
+
+The Job Controller filters all pods and classifies the pods as active.  This list is used to verify that the expected number of active pods matches the reality.  
+
+As part of this KEP, we want to include pods that are also terminating (`DeletionTimestamp != nil`).  
+The field `Status.terminating` will include the number of terminating pods.  Once the pods are fully terminated, this field will be unset again because we will obtain the count of active and terminating pods and update the status.
+
+We will update the Status field in the same place as when we update the number of active jobs so it should not require any extra API calls.  
+
+In cases where pods are terminating outside of the job controller, we will look at the list of pods that a job has.  And we can classify them as terminating and add that to the status field.
+
+### Test Plan
+
+<!--
+**Note:** *Not required until targeted at a release.*
+The goal is to ensure that we don't accept enhancements with inadequate testing.
+
+All code is expected to have adequate tests (eventually with coverage
+expectations). Please adhere to the [Kubernetes testing guidelines][testing-guidelines]
+when drafting this test plan.
+
+[testing-guidelines]: https://git.k8s.io/community/contributors/devel/sig-testing/testing.md
+-->
+
+[x] I/we understand the owners of the involved components may require updates to
+existing tests to make this code solid enough prior to committing the changes necessary
+to implement this enhancement.
+
+#### Prerequisite testing updates
+
+<!--
+Based on reviewers feedback describe what additional tests need to be added prior
+implementing this enhancement to ensure the enhancements have also solid foundations.
+-->
+
+##### Unit tests
+
+- `controller_utils`: `April 3rd 2023` - `56.6`
+- `job`: `April 3rd 2023` - `90.4`
```

Comment:
What unit tests are you planning to add to cover the new functionality?

Reply:
I tried to add some examples of the tests.  In some cases I already have an implementation in mind but I need to look more into gc_controller and understand if I should have explicit integration tests around this in addition to unit tests.  

I think that this shouldn't hold the KEP review though as that seems to be more of an implementation detail.